### PR TITLE
[Feature]Add Residual Log-likelihood Estimation(RLE) as a new loss function

### PR DESCRIPTION
**Pull request by Indigo6**
## Motivation

RLE is a flow-based adaptive regression loss proposed in ICCV2021. It estimates the underlying distribution of error by maximizing the likelihood of errores appearing during training.

Arch | Input Size | Mean | Mean@0.1 |
-- | -- | -- | -- | 
deeppose_resnet_50 | 256x256 | 0.825 | 0.174 
deeppose_resnet_50_rle | 256x256 | 0.866 | 0.273

> Original Paper: [Human Pose Regression With Residual Log-Likelihood Estimation](https://openaccess.thecvf.com/content/ICCV2021/html/Li_Human_Pose_Regression_With_Residual_Log-Likelihood_Estimation_ICCV_2021_paper.html)
> [Original Code](https://github.com/Jeff-sjtu/res-loglikelihood-regression)

</body>
</html>

## Modification

1. Add RLELoss in regression_loss
2. Add `realnvp` in `models/utils/`
3. Add `out_sigma` arg/opt in deepose_regression_head to predict $\hat\sigma$
4. Add deepose configs using RLELoss

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.
[deeppose_resnet_50](https://github.com/open-mmlab/mmpose/tree/master/configs/body/2d_kpt_sview_rgb_img/deeppose/mpii/res50_mpii_256x256.py)	256x256	0.825	0.174

### CODE REVIEW

**Comment by jin-s13** [comment]
Thank you for your contribution! We appreciate it very much. We will review this PR ASAP.

**Comment by ly015** [comment]
@Indigo6 Thank you very much for your contribution. We would like to know if you plan to share the pretrained models so they can be released in MMPose Model Zoo. Or we can train the models following the configs.

**Comment by Indigo6** [comment]
> @Indigo6 Thank you very much for your contribution. We would like to know if you plan to share the pretrained models so they can be released in MMPose Model Zoo. Or we can train the models following the configs.

For the MPII dataset, both are ok to me as the pretrained model and training log with `res50_mpii_256x256_rle.py` are already done. For COCO, I prefer leaving it to you since it takes a long time to train on my single-gpu device.

**Comment by ly015** [comment]
> > @Indigo6 Thank you very much for your contribution. We would like to know if you plan to share the pretrained models so they can be released in MMPose Model Zoo. Or we can train the models following the configs.
> 
> @ly015 For the MPII dataset, both are ok to me as the pretrained model and training log with `res50_mpii_256x256_rle.py` are already done. For COCO, I prefer leaving it to you since it takes a long time to train on my single-gpu device.

Thank you, that would be great. Could you please provide a link so we can access the pretrained model on MPII? And we will train the COCO model.

**Comment by Indigo6** [comment]
> > > @Indigo6 Thank you very much for your contribution. We would like to know if you plan to share the pretrained models so they can be released in MMPose Model Zoo. Or we can train the models following the configs.
> > 
> > 
> > @ly015 For the MPII dataset, both are ok to me as the pretrained model and training log with `res50_mpii_256x256_rle.py` are already done. For COCO, I prefer leaving it to you since it takes a long time to train on my single-gpu device.
> 
> Thank you, that would be great. Could you please provide a link so we can access the pretrained model on MPII? And we will train the COCO model.

[Google Drive](https://drive.google.com/drive/folders/1UuotKQipcDvNGdZ0ldYcwpU2hpWX4W0j?usp=sharing)
One thing to notice: I set the `interval` of evaluation to `1` when training. I can upload that of `interval=10`, but it may take 10 hours.

File: mmpose/models/utils/realnvp.py [hunk 1]
```diff
@@ -18,11 +18,12 @@ def __init__(self, get_scale_net, get_trans_net, mask, prior):
         super(RealNVP, self).__init__()
 
         self.prior = prior
-        self.register_buffer('mask', mask)
+        self.mask = mask
```
Comment (Jeff-sjtu) [inline mmpose/models/utils/realnvp.py:21]: Is there a problem with not registering the mask in the buffer? It seems the mask won't be stored in the model state dict.

Comment (Indigo6) [reply mmpose/models/utils/realnvp.py:21]: > Is there a problem with not registering the mask in the buffer? It seems the mask won't be stored in the model state dict.

The flow model won't be stored in the model state dict since I implemented RLE as a loss function.
For this line, I modified it to unify the devices of the variables of the flow model. When adding unit test and trying out, the device of `mask` will be different from`prior` using `register_buffer`.

Comment (Jeff-sjtu) [reply mmpose/models/utils/realnvp.py:21]: The device of the buffer can change with the module by simply calling `loss.cuda()`, or `loss.to(device)`, since the buffer is registered in the module and its device will stay the same as the parameters. You only need to move the device of the `prior` to GPU.

Besides, the loss module is registered in the regression head. I think the parameters of the flow model will resume with the regression head.

Comment (Indigo6) [reply mmpose/models/utils/realnvp.py]: You are right, I loaded the checkpoint, loss is a variable of model.head and its parameters are stored. However, it just came to my mind that should `mask` be register&stored? I remember it's a constant?
Another thing, there are some parameters that can't be personalized right now, like `prior`, `get_scale_net(nets in your code)`, `get_trans_net(nett in your code)`, and `Q`. I'm not sure whether to add options for them or not.

**Comment by Indigo6** [comment]
I read the error report of build, it seems to be the `gpu_test` for RLELoss. I'll remove it in the next commit. This comment is for asking if removing it is ok since there are some codes for transfering `RLELoss.prior` to GPU(s)

**Comment by ly015** [comment]
> 

@Indigo6 It's ok for now. Please make sure that the unit test can cover the major part of the code. And we will also help check the reason that causes CI failure with CUDA.

File: mmpose/models/heads/deeppose_regression_head.py [hunk 2]
```diff
@@ -26,7 +26,8 @@ def __init__(self,
                  num_joints,
                  loss_keypoint=None,
                  train_cfg=None,
-                 test_cfg=None):
+                 test_cfg=None,
+                 out_sigma=False):
```
Comment (ly015) [inline mmpose/models/heads/deeppose_regression_head.py:30]: I suggest placing `out_sigma` before `train_cfg`. It should not introduce any bc-breaking.

Comment (Indigo6) [reply mmpose/models/heads/deeppose_regression_head.py:30]: Right, splitting `args`  and `cfg_args` is better

Comment (ly015) [reply mmpose/models/heads/deeppose_regression_head.py:30]: I will do it when I finish the review :)

Comment (Indigo6) [reply mmpose/models/heads/deeppose_regression_head.py:30]: Ok, thank you~

File: mmpose/models/utils/realnvp.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+import torch.nn as nn
+
+
+class RealNVP(nn.Module):
```
Comment (ly015) [inline mmpose/models/utils/realnvp.py:6]: Is the implementation of RealNVP adapted from other projects? If so, we may need to add license and attribution information properly.

File: =1.3.17 [hunk 4]
```diff
@@ -0,0 +1,11 @@
+Looking in indexes: https://pypi.tuna.tsinghua.edu.cn/simple
```
Comment (ly015) [inline =1.3.17:1]: This file seems irrelevant.

Comment (Indigo6) [reply =1.3.17:1]: Yes... This is weired. I tried the coverage test on my device, and it reported "need MMCV>=3.1.17".
It may appear when I reinstall MMCV. I'll delete it

**Comment by Indigo6** [comment]
@ly015 What am I supposed to do to pass the codecov/project check? As the codecov/patch check shows, all of my modifications are covered by unit tests.

File: mmpose/models/utils/realnvp.py (hunk 1 above)
Comment (Indigo6) [reply mmpose/models/utils/realnvp.py:21]: Loading mask and prior to GPU are solved in commit [fix device bug of flow by overriding loss._apply(fn)](https://github.com/open-mmlab/mmpose/pull/1259/commits/76e78cd539b490e207c428deb495e3d641cc009e)

**Comment by ly015** [comment]
@Indigo6 Sorry to bother you, but is there any update to the unsolved comments? We are looking forward to your feedback so we can help organize the code :)

**Comment by Indigo6** [comment]
> @Indigo6 Sorry to bother you, but is there any update to the unsolved comments? We are looking forward to your feedback so we can help organize the code :)

All reviews and comments are resolved in recent commits. 
For the latest unresolved review, I don't know whether and how to "add license and attribution information", and was waiting for your reply.
All in all, I think the codes are ready for review.

**Comment by ly015** [comment]
> > @Indigo6 Sorry to bother you, but is there any update to the unsolved comments? We are looking forward to your feedback so we can help organize the code :)
> 
> All reviews and comments are resolved in recent commits. For the latest unresolved review, I don't know whether and how to "add license and attribution information", and was waiting for your reply. All in all, I think the codes are ready for review.

Thanks for your reply. We will add license information later.

I had a few questions about the RLE loss implementation and some thoughts about the class design, per previous comments. It would be great if you can share some options.

**Comment by Indigo6** [comment]
> > > @Indigo6 Sorry to bother you, but is there any update to the unsolved comments? We are looking forward to your feedback so we can help organize the code :)
> > 
> > 
> > All reviews and comments are resolved in recent commits. For the latest unresolved review, I don't know whether and how to "add license and attribution information", and was waiting for your reply. All in all, I think the codes are ready for review.
> 
> Thanks for your reply. We will add license information later.
> 
> I had a few questions about the RLE loss implementation and some thoughts about the class design, per previous comments. It would be great if you can share some options.

@ly015 What should I do to "share some options"? Do you mean re-open conversations?
Feel free to ask anything, I'm a second-year graduate student and when not worrying about the future I have plenty of free time.
Thank you for your excellent work and patient review.

**Comment by ly015** [comment]
> > > > @Indigo6 Sorry to bother you, but is there any update to the unsolved comments? We are looking forward to your feedback so we can help organize the code :)
> > > 
> > > 
> > > All reviews and comments are resolved in recent commits. For the latest unresolved review, I don't know whether and how to "add license and attribution information", and was waiting for your reply. All in all, I think the codes are ready for review.
> > 
> > 
> > Thanks for your reply. We will add license information later.
> > I had a few questions about the RLE loss implementation and some thoughts about the class design, per previous comments. It would be great if you can share some options.
> 
> @ly015 What should I do to "share some options"? Do you mean re-open conversations? Feel free to ask anything and thank you for your excellent work & patient review. I'm a second-year graduate student and when not worrying about the future I have plenty of free time.

@Indigo6 Oh, sorry that these comments were not submitted so they were not visible to you. That's a little embarrassing... Anyway, I just submitted them. Please take a look at your convenience. Thank you!

File: mmpose/models/losses/regression_loss.py [hunk 5]
```diff
@@ -4,8 +4,107 @@
...
+            target_weight (torch.Tensor[N, K, D]):
+                Weights across different joint types.
+        """
+        sigma = output[:, :, 2:].sigmoid()
+        coord = output[:, :, :2]
+
+        error = (coord - target) / sigma
+        # (B, K, 2)
+        log_phi = self.flow.log_prob(error.reshape(-1, 2))
+        log_phi = log_phi.reshape(target.shape[0], target.shape[1], 1)
```
[the remark this answers is not in the document]
Comment (Indigo6) [reply mmpose/models/losses/regression_loss.py:87]: As far as I know, the flow accepts input of `(N_F, 2)` shape, where `N_F` stands for the batch_size of the flow.
Therefore, the results need to be reshaped back after merging dim=0 and dim=1.

File: mmpose/models/losses/regression_loss.py [hunk 6]
```diff
@@ -4,8 +4,107 @@
...
+        super(RLELoss, self).__init__()
+        self.size_average = size_average
+        self.use_target_weight = use_target_weight
+        self.residual = residual
+        self.amp = 1 / math.sqrt(2 * math.pi)
+
+        prior = distributions.MultivariateNormal(torch.zeros(2), torch.eye(2))
+        masks = torch.tensor([[0, 1], [1, 0]] * 3, dtype=torch.float32)
+        self.flow = RealNVP(self.get_scale_net, self.get_trans_net, masks,
+                            prior)
```
[the remark this answers is not in the document]
Comment (Indigo6) [reply mmpose/models/losses/regression_loss.py:51]: You are right, it's better.

[the remark this answers is not in the document]
Comment (Indigo6) [reply mmpose/models/losses/regression_loss.py:51]: What's your opinion about adding options for customing `prior` and `s/t`?

File: mmpose/models/losses/regression_loss.py [hunk 7]
```diff
@@ -4,8 +4,107 @@
...
+                                             2)
+        nf_loss = log_sigma - log_phi
+
+        if self.residual:
+            loss_sigma = torch.log(sigma / self.amp)
+            loss_l1 = torch.abs(target - coord)
+            loss_l1 /= (math.sqrt(2) * sigma + 1e-9)
+            loss = nf_loss + loss_l1 + loss_sigma
+        else:
+            loss = nf_loss
```
[the remark this answers is not in the document]
Comment (Indigo6) [reply mmpose/models/losses/regression_loss.py:98]: I need to re-read the paper and original project for these questions. Is it ok to reply after some days?

File: mmpose/models/losses/regression_loss.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ly015) [reply mmpose/models/losses/regression_loss.py:51]: I think we can focus on supporting the default settings (as in the paper) of `prior` and `s/t` in this PR. It should be easy to extend to more options in the future.
Of course, if you already have some checkpoints or experiment results using other settings, we can add them to this PR.

File: mmpose/models/losses/regression_loss.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ly015) [reply mmpose/models/losses/regression_loss.py:98]: No problem, take your time, please. Thanks for your help!

[the remark this answers is not in the document]
Comment (Indigo6) [reply mmpose/models/losses/regression_loss.py:98]: Sorry for the late reply, please refer to the attached file since in-line formula is not supported by Github.
[Reply.md](https://github.com/open-mmlab/mmpose/files/8424900/Reply.md)
[Reply.pdf](https://github.com/open-mmlab/mmpose/files/8424901/Reply.pdf)

File: mmpose/models/losses/regression_loss.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Indigo6) [reply mmpose/models/losses/regression_loss.py]: Done in commit [7fe7202](https://github.com/open-mmlab/mmpose/pull/1259/commits/7fe720290a49275450338b64e88752797a309fd1)

File: mmpose/models/losses/regression_loss.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Indigo6) [reply mmpose/models/losses/regression_loss.py]: I simplified the implementation of log(Q) in commit [facb0d0](https://github.com/open-mmlab/mmpose/pull/1259/commits/facb0d01388b3216668ba3edb4bd3346c50c0bc6). 
Experiments are going on to check the influence.

[the remark this answers is not in the document]
Comment (Indigo6) [reply mmpose/models/losses/regression_loss.py]: > I simplified the implementation of log(Q) in commit [facb0d0](https://github.com/open-mmlab/mmpose/pull/1259/commits/facb0d01388b3216668ba3edb4bd3346c50c0bc6). Experiments are going on to check the influence.

The experimental result is 86.2, 0.4 lower than the original implementation··· I am confused

[the remark this answers is not in the document]
Comment (Indigo6) [reply mmpose/models/losses/regression_loss.py]: Sorry for the late reply.
First of all, the forward process of flow is:
$$
\bar\mu=\phi(z)\\
\mu = \bar\mu*\hat\sigma+\hat\mu
$$
where $\hat\mu$ and $\mu$ is the predicted and gt joints, $\hat\sigma$ is the predicted variance. 

1. The `error` in my code is $\bar\mu$ in the paper, which is:
   $$
   \bar\mu=f^{-1}(\mu)=\frac{\hat\mu-\mu}{\hat\sigma}
   $$
   We know that when  $y=f(x)$, we have:
   $$
   logP_f(y)=logP(x)+log|det(J(f^{-1}))|
   $$
   Thus, the loss without residual is:
   $$
   -logP(\mu)=-logP_\phi(\bar\mu)+log\ \hat\sigma
   $$
   where `self.flow.log_prob(error)` is $-logP(\bar\mu)$

2. For the ***double log(sigma)***, I think the author may mis-implemented the loss from that in the paper:
   $$
   -logP(\mu)=-logP(\bar\mu)+log\ \hat\sigma\\
   log_\phi P(\bar\mu) = logQ(\bar\mu)+log_\phi G(\bar\mu) + log s
   $$
   to:
   $$
   -logP(\mu)=-logP(\bar\mu)-logQ(\mu)+log\ \hat\sigma\\
   log_\phi P(\bar\mu) = log_\phi G(\bar\mu) + log s
   $$
   In the first formula, the variance for the distribution of $\bar\mu$ should be constant not $\hat\sigma$ since it was already re-scaled.

3. Yes, it's ignored, you can refer to the appendix of the paper

   > The term s barely affects the results and can be removed to save computation resources. Therefore, in our implementation, we drop the term log s for simplicity.

4. If I am right in the ***double log(sigma)*** question, I think it's unnecessary for this constant factor anymore.

[the remark this answers is not in the document]
Comment (Indigo6) [reply mmpose/models/losses/regression_loss.py]: I simplified the implementation of log(Q) in commit [facb0d0](https://github.com/open-mmlab/mmpose/pull/1259/commits/facb0d01388b3216668ba3edb4bd3346c50c0bc6).

[the remark this answers is not in the document]
Comment (Indigo6) [reply mmpose/models/losses/regression_loss.py]: @ly015 Sorry to bother, just wondering if you've seen my reply since I didn't @ you

[the remark this answers is not in the document]
Comment (ly015) [reply mmpose/models/losses/regression_loss.py]: Yes. Sorry for the late reply. I will review the latest updates this week. Thank you very much.

[the remark this answers is not in the document]
Comment (Indigo6) [reply mmpose/models/losses/regression_loss.py]: Ok! Thank you for your reply and review :wink: !

**Comment by jin-s13** [comment]
This PR looks good to me. We will retrain a few models to check the accuracy before merging.

**Comment by kennymckormick** [comment]
Hi, @Indigo6 , will you also add a config for the CoCo dataset?

File: mmpose/models/losses/regression_loss.py [hunk 8]
```diff
@@ -6,6 +6,94 @@
...
+
+        if self.use_target_weight:
+            assert target_weight is not None
+            loss *= target_weight
+
+        if self.size_average:
+            loss /= len(loss)
+
+        return loss.sum()
```
Comment (ly015) [inline mmpose/models/losses/regression_loss.py:96]: @Indigo6 Thanks for the detailed [explanation](https://github.com/open-mmlab/mmpose/pull/1259#discussion_r844574413) of the loss implementation. I am clear about Q2-Q4. 

About Q1, as you explained, `self.flow.log_prob(error)` is $-logP(\bar\mu)$. However in [`RealNVP.log_prob()`](https://github.com/open-mmlab/mmpose/pull/1259/files#diff-711f9876f89558df68e9c40a354cd40316a556a3922b998b0f8f2233494d95daR57-R59) it seems that the return value is the summation of $logP(\bar\mu)$ and $log_det$. The second term, according to Eq.5 in the paper, should be $log\ \hat\sigma$, which leads to my previous question that `log_phi` seems to already contain the $log\ \hat\sigma$ term and why it's added again in [L77](https://github.com/open-mmlab/mmpose/pull/1259/files#diff-9674ca04b72898ff995646744e5d26d5a8b6346de33d2c564d5f45ed9467e6b6R77). Please correct me if I misunderstand the math somewhere.

The rest parts look all good to me and we will very the model training this week. And if you don't mind, we may probably make some minor modifications and directly push them to this PR. Feel free to leave your comments and suggestions on these changes.

**Comment by Indigo6** [comment]
> Hi, @Indigo6 , will you also add a config for the CoCo dataset?

Sure, and you just need to change the model['keypoint_head']['loss_keypoint'] to the following for using RLE.
> loss_keypoint=dict(
            type='RLELoss',
            use_target_weight=True,
            size_average=True,
            residual=True),

File: mmpose/models/losses/regression_loss.py (hunk 8 above)
Comment (Indigo6) [reply mmpose/models/losses/regression_loss.py:96]: > @Indigo6 Thanks for the detailed [explanation](https://github.com/open-mmlab/mmpose/pull/1259#discussion_r844574413) of the loss implementation. I am clear about Q2-Q4.
> 
> About Q1, as you explained, `self.flow.log_prob(error)` is −logP(μ¯)-logP(\bar\mu). However in [`RealNVP.log_prob()`](https://github.com/open-mmlab/mmpose/pull/1259/files#diff-711f9876f89558df68e9c40a354cd40316a556a3922b998b0f8f2233494d95daR57-R59) it seems that the return value is the summation of logP(μ¯)logP(\bar\mu) and logdetlog_det. The second term, according to Eq.5 in the paper, should be log σ^log\ \hat\sigma, which leads to my previous question that `log_phi` seems to already contain the log σ^log\ \hat\sigma term and why it's added again in [L77](https://github.com/open-mmlab/mmpose/pull/1259/files#diff-9674ca04b72898ff995646744e5d26d5a8b6346de33d2c564d5f45ed9467e6b6R77). Please correct me if I misunderstand the math somewhere.
> 
> The rest parts look all good to me and we will very the model training this week. And if you don't mind, we may probably make some minor modifications and directly push them to this PR. Feel free to leave your comments and suggestions on these changes.

@ly015 Hi, about Q1, there are two functions $f$ and $\phi$, and two corresponding $log|det(J(f^{-1}))|$ and $log|det(J(\phi^{-1}))|$. The first function stands for the reparameterization: 
$$
\mu = f(\bar\mu) = \bar\mu*\hat\sigma+\hat\mu
$$ where its $log|det(J(f^{-1}))|$ is $log \hat\sigma$.\\
The second function is the RealNVP:
$$
\bar\mu=\phi(z)
$$ where its $log|det(J(\phi^{-1}))|$ is not $log \hat\sigma$. In other words, `self.flow.log_prob(error)` do not include $log \hat\sigma$. \\

About model performance, it may be a little lower than the original implementation. After I simplified the implementation of log(Q) in commit [facb0d0](https://github.com/open-mmlab/mmpose/pull/1259/commits/facb0d01388b3216668ba3edb4bd3346c50c0bc6), I retrained the model and the performance degraded a bit. 

About direct modifications, it's ok for me.

Comment (ly015) [reply mmpose/models/losses/regression_loss.py:96]: Thanks for your reply. About the performance degradation, do you think it's caused by the removal of `loss_sigma`?

Comment (Indigo6) [reply mmpose/models/losses/regression_loss.py:96]: Yes, the removal of `loss_sigma` should be the reason, but I don't know why it caused performance degradation
